core/lib/contract_verifier/src/lib.rs

```diff
@@ -353,6 +364,99 @@ impl ContractVerifier {
         Ok((info, compiled_identifier))
     }
 
+    /// Returns compilation artifacts for the given request.
+    /// Tries to use the compiler versions from the metadata if they are different from the ones in the request.
+    /// If the metadata is not present, or the versions are the same, uses the ones from the request.
+    async fn get_compilation_artifacts(
+        &self,
+        request: &mut VerificationRequest,
+        detected_metadata: Option<&DetectedMetadata>,
+        bytecode_marker: &BytecodeMarker,
+        deployed_code: &[u8],
+    ) -> Result<CompilationArtifacts, ContractVerifierError> {
+        match detected_metadata {
```

Comment:
There is a bit too much nesting in this function. I guess it can be simplified to something like this (pseudocode):

```rust
// If compiler versions from the metadata don't match with the request,
// we'll first try to use info from metadata.
let alternative_request = detected_metadata
  .filter_map(|metadata| {
    if !request.compiler_versions_match(&metadata) {
      let updated_request = ...;
      Some(updated_request)
    }  else {
       None
    });
if let Some(alternative_request) = alternative_request {
    // compile
    // check against identifier
    // if matches, save it and return artifacts
}

// We either have no alternative request or it didn't succeed
self.compile(request.req.clone(), *bytecode_marker).await
```

Reply:
refactored `get_compilation_artifacts` fn, please check